service/accounts/openstack_manager.py

```diff
@@ -253,8 +303,12 @@ def get_or_create_keypair(self, username, password, project_name,
         public_key - Contents of public key in OpenSSH format
         """
         clients = self.get_openstack_clients(username, password, project_name)
-        nova = clients["nova"]
-        keypairs = nova.keypairs.list()
+        if self.identity_version == 2:
+            nova = clients["nova"]
+            keypairs = nova.keypairs.list()
+        else:
+            osdk = clients["openstack_sdk"]
+            keypairs = [kp for kp in osdk.compute.keypairs()]
```

Comment:
Okay - this is the _"either or"_ pattern you mentioned to me earlier. Got it :+1: